src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs

```diff
@@ -111,33 +112,65 @@ public async Task StartAsync(string target = null)
         {
             // Codes in try block should catch and log exceptions separately.
             // The catch here should be very rare to reach.
-            try
+            // Don't allow a second call to StartAsync
+            if (Status != ServiceConnectionStatus.Inited)
+            {
+                throw new InvalidOperationException("Connection already started!");
+            }
+
+            Status = ServiceConnectionStatus.Connecting;
+            var connection = await EstablishConnectionAsync(target);
+            if (connection != null)
             {
-                if (await StartAsyncCore(target))
+                _connectionContext = connection;
+                Status = ServiceConnectionStatus.Connected;
+                _serviceConnectionStartTcs.TrySetResult(true);
+                try
                 {
-                    _serviceConnectionStartTcs.TrySetResult(true);
-                    await ProcessIncomingAsync();
+                    await ProcessIncomingAsync(connection);
+
+                    // when ProcessIncoming completes, clean up the connection
+
+                    // TODO: Never cleanup connections unless Service asks us to do that
+                    // Current implementation is based on assumption that Service will drop clients
+                    // if server connection fails.
+                    await CleanupConnections();
                 }
+                catch (Exception ex)
+                {
+                    Log.ConnectionDropped(Logger, HubEndpoint.ToString(), ConnectionId, ex);
+                }
+                finally
+                {
+                    // wait until all the connections are cleaned up to close the outgoing pipe
+                    // mark the status as Disconnected so that no one will write to this connection anymore
+                    await _writeLock.WaitAsync();
+                    try
+                    {
+                        // Don't allow write anymore when the connection is disconnected
+                        Status = ServiceConnectionStatus.Disconnected;
 
-                _serviceConnectionStartTcs.TrySetResult(false);
+                        // close the underlying connection
+                        await DisposeConnection(connection);
+                    }
+                    finally
+                    {
+                        _writeLock.Release();
+                    }
+                }
             }
-            catch (Exception ex)
+            else
             {
+                _serviceConnectionStartTcs.TrySetResult(false);
```

Comment:
Do we even use the boolean flag anywhere?

Reply:
no, but I feel OK to keep it this way?

---
In reply to: [334723761](https://github.com/Azure/azure-signalr/pull/672#discussion_r334723761) [](ancestors = 334723761)